app/com/linkedin/drelephant/tuning/hbt/PigHbtParameterRecommender.java

```diff
@@ -0,0 +1,406 @@
+          for (AppHeuristicResultDetails heuristicResultDetail : appHeuristicResult.yarnAppHeuristicResultDetails) {
+            if (heuristicResultDetail.name.equals(MAPPER_MEMORY_HEURISTICS_CONF.getValue())) {
+              latestAppliedParams.put(MAPPER_MEMORY_HADOOP_CONF.getValue(),
+                  Double.parseDouble(heuristicResultDetail.value));
+            } else if (heuristicResultDetail.name.equals(REDUCER_MEMORY_HEURISTICS_CONF.getValue())) {
+              latestAppliedParams.put(REDUCER_MEMORY_HADOOP_CONF.getValue(),
+                  Double.parseDouble(heuristicResultDetail.value));
+            } else if (heuristicResultDetail.name.equals(SORT_BUFFER_HEURISTICS_CONF.getValue())) {
+              latestAppliedParams.put(SORT_BUFFER_HADOOP_CONF.getValue(),
+                  Double.parseDouble(heuristicResultDetail.value));
+            } else if (heuristicResultDetail.name.equals(SORT_FACTOR_HEURISTICS_CONF.getValue())) {
+              latestAppliedParams.put(SORT_FACTOR_HADOOP_CONF.getValue(),
+                  Double.parseDouble(heuristicResultDetail.value));
+            } else if (heuristicResultDetail.name.equals(PIG_MAX_SPLIT_SIZE_HEURISTICS_CONF.getValue())) {
+              latestAppliedParams.put(PIG_SPLIT_SIZE_HADOOP_CONF.getValue(),
+                  Double.parseDouble(heuristicResultDetail.value));
+            } else if (heuristicResultDetail.name.equals(SORT_SPILL_HEURISTICS_CONF.getValue())) {
+              latestAppliedParams.put(SORT_SPILL_HADOOP_CONF.getValue(),
+                  Double.parseDouble(heuristicResultDetail.value));
+            } else if (heuristicResultDetail.name.equals(MAPPER_HEAP_HEURISTICS_CONF.getValue())) {
+              latestAppliedParams.put(MAPPER_HEAP_HADOOP_CONF.getValue(), getHeapMemory(heuristicResultDetail.value));
+            } else if (heuristicResultDetail.name.equals(REDUCER_HEAP_HEURISTICS_CONF.getValue())) {
+              latestAppliedParams.put(REDUCER_HEAP_HADOOP_CONF.getValue(), getHeapMemory(heuristicResultDetail.value));
+            }
+          }
+        }
+      }
+    } else {
+      throw new IllegalArgumentException("No App Results found while TuneIn was suggesting new parameters");
+    }
+  }
+
+  /**
+   * This method is responsible for suggesting the parameters to fix failed heuristics
+   * in the next execution or if there is no failed heuristics then try to optimize Resource usage
+   * @return Map with suggested parameter's name as key and its value
+   */
+  public Map<String, Double> suggestParameters() {
+    loadLatestAppliedParameters();
+    List<String> failedHeuristicNameList = getFailedHeuristics();
+    suggestToFixFailedHeuristics(failedHeuristicNameList);
+    if (failedHeuristicNameList.size() == 0) {
+      logger.info("No failed heuristics, so will try to optimize memory parameters");
+      suggestMemoryParam(MRJobTaskType.MAP);
+      suggestMemoryParam(MRJobTaskType.REDUCE);
+    }
+    addRequiredPreviousSuggestedParameters();
```

Comment:
Shouldn't this be called after modifyMapperMemory ?

Reply:
https://github.com/linkedin/dr-elephant/pull/608#discussion_r300560725